### ISSUE TITLE
Local variable referenced in trap handler with global scope

### ISSUE THREAD

**Issue by fb913bf0de288ba84fe98f7a23d35edfdb22381** [koalaman/shellcheck#1932]
#### For new checks and feature suggestions
- [x ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

function handler() {
        echo "Handler: ${device_name}"
}

function do_something() {
        local device_name="abc"
        trap handler EXIT
}

do_something
```

Output:

```
$ bash test.sh 
Handler: 
```

#### Here's what shellcheck currently says:

`No issues detected!`

#### Here's what I wanted or expected to see:

Warning that the local variable referenced in the handler is out of scope when the handler is invoked.

Note that for the RETURN trap handler, the variable would still be in scope. No warning should be needed there. The warning should only apply for signals that may be invoked after the scope of the variables referenced in their handlers.

**Comment by fb913bf0de288ba84fe98f7a23d35edfdb22381** [comment]
I now realize there seems to be a more general issue in shellcheck with the detection of unassigned variables.

```
#!/bin/bash

f1() {
        echo "myvar=${myvar}"
}

f2() {
        local myvar="abc"
}

f1
```
Expected Output: Warning about myvar being referenced before assignment.

Actual Output:
```
$ bash test.sh 
myvar=
```
Another example:
```
f2() {
        echo "myvar=${myvar}"
        local myvar="abc"
}

f2
```
Expected vs. Actual Output as above.

If I understand the current code correctly ([link](https://github.com/koalaman/shellcheck/blob/master/src/ShellCheck/Analytics.hs#L2231)), it does a diff on the variables that are read vs the variables that are assigned, without considering wether the assignment happens before the read? I.e. it does not take the actual control flow in consideration. This would probably have to be solved first before taking care of the trap issues.
